Update the pull request template (#1165)
chaos-polymtl/lethe: CONTRIBUTING.md

# Lethe

Lethe is an open source software, and we try to follow the deal.II mentality of having it
developed in an accessible environment. Consequently, all scientific

> developments made within Lethe are accessible to everyone even as they are

Reviewer: ```suggestion
developments made within Lethe are available to everyone even as they are
```

currently being written. Anybody that wishes to use and contribute to a feature

> is welcome to do so.

Reviewer: ```suggestion
is welcomed to do so.
```

This document aims at establishing guidelines for contributions within Lethe.

> Lethe has its own documentation (user guide and developer guide) which are openly accessible on: https://chaos-polymtl.github.io/lethe/index.html

Reviewer 2: ```suggestion
Lethe has its own documentation (user guide and developer guide) which are available on: https://chaos-polymtl.github.io/lethe/index.html
```

Author: Bruno suggested to change accessible to available, will let you fight on that one 👀

This document aims at establishing guidelines for contributions within
Lethe.

> Lethe has a wiki which is openly accessible on : https://chaos-polymtl.github.io/lethe/index.html

Author: I agree it does not anymore

request at the official Github repository of Lethe : https://github.com/chaos-polymtl/lethe.
We recommend that users either request access to the Lethe repository or
create their own fork of Lethe on their own Github account. This can then be
used to open pull requests

# What should be the content of a pull request (PR)?

A pull request should contain the following elements:

- A brief title (less than 60 characters) that describes the goal of the pull
request.
- A detailed description of the content added by the pull request:
  * If the PR adds a new feature:\
    The feature should be documented, tested (with unit tests and/or application tests) 
    and if the feature adds or alters parameters in the input file, the appropriate section 
    of the documentation should be updated accordingly.
  * If the pull request corrects a bug:\
    the source of the bug should be explained and the way it was identified should be briefly described. 
    A unit test or application test that reproduces the bug should be added.

See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

> 

Reviewer 2: Unrelated:
This is maybe not for this PR and should be added to the "Miscellaneous" subsection of the PR, but a check should be done to make sure that all information here correspond to what's on the online documentation and vice-versa.
For example, the 60 char rule is not mentioned in the online doc and in this file the `prmindent` is not mentioned. However, since you link the doc here, you don't need to put all the info that's there here, but there is no link from the doc to here so the other way is the still a problem.

Author: Actually, if the info is somewhere, it should not be somewhere else also. Meaning, we should refer to the doc only. 
Let see what Bruno has to say about that

Reviewer 2: I remember that there was an issue regarding the fact that when the documentation was down (during the time doxygen was added) it was not possible to know how to locally compile from the documentation online. Other than that (how to locally compile documentation), I think you're right and we could leave the rest online.

# Good practices

> - Before making a pull request, the branch should be rebased on master to ensure

Reviewer 2: ```suggestion
- Before making a pull request, the branch should be rebased onto master to ensure
```

a linear history and to make merging as easy as possible. It is the responsibility
of the branch owner to ensure that the pull request goes as smoothly as possible.
Pushing to Github after a rebase requires a force push. You can accomplish
this by using `git push --force-with-lease`

> - All functions should be documented in the `.h`. All functions should contain
> a `@brief` that describes the function and a `@param` for each argument.

Reviewer 2: Unrelated:
```suggestion
- All functions should be documented in the `.h`. All functions should contain
a `@brief` that describes the function, a `@param` for each argument, and a `@return` if anything is returned by the function.
```

Author: I did not know we needed the @return actually. I guess our "official" Doxygen rules should be all listed in the contributing doc and not here

those libraries and executables) in the `CMakeLists.txt` files'
`TARGET_LINK_LIBRARIES` calls, and omit any transitive dependencies.

For example, the `lethe-particles` application
[links to `lethe-core` and `lethe-dem`](applications/lethe-particles/CMakeLists.txt)
because it
[includes core and DEM headers](applications/lethe-particles/dem_2d.cc),
but the `lethe-fluid` application
[links only to `lethe-solvers`](applications/lethe-fluid/CMakeLists.txt)
because it

> [includes only solvers headers](applications/lethe-fluid/gd_navier_stokes_2d.cc),

Reviewer: We should put the name of the executable as they are now here. dem_2d and gd_navier_stokes_2d do not exist anymore. :)

those libraries and executables) in the `CMakeLists.txt` files'
`TARGET_LINK_LIBRARIES` calls, and omit any transitive dependencies.

For example, the `lethe-particles` application

> [links to `lethe-core` and `lethe-dem`](applications/dem_2d/CMakeLists.txt)

Reviewer 2: ```suggestion
[links to `lethe-core` and `lethe-dem`](applications/lethe-particles/CMakeLists.txt)
```

because it

> [includes core and DEM headers](applications/dem_2d/dem_2d.cc),

Reviewer 2: ```suggestion
[includes core and DEM headers](applications/lethe-particles/dem.cc),
```

but the `lethe-fluid` application

> [links only to `lethe-solvers`](applications/gd_navier_stokes_2d/CMakeLists.txt)

Reviewer 2: ```suggestion
[links only to `lethe-solvers`](applications/lethe-fluid/CMakeLists.txt)
```

because it

> [includes only solvers headers](applications/gd_navier_stokes_2d/gd_navier_stokes_2d.cc),

Reviewer 2: ```suggestion
[includes only solvers headers](applications/lethe-fluid/gls_navier_stokes.cc),
```
